# Recommendation Systems


 Recommendation Systems are a type of information filtering systems as they improve the quality of search results and provides items that are more relevant to the search item or are realted to the search history of the user.

## Demographic Filtering

They offer generalized recommendations to every user, based on movie popularity and/or genre. The System recommends the same movies to users with similar demographic features. Since each user is different , this approach is considered to be too simple. The basic idea behind this system is that movies that are more popular and critically acclaimed will have a higher probability of being liked by the average audience.

#### Let's load the data now

In [1]:
import numpy as np
import pandas as pd
df1=pd.read_csv('tmdb_5000_credits.csv')
df2=pd.read_csv('tmdb_5000_movies.csv')

#### The dataset is as follows

The first dataset contains the following features:-

    movie_id - A unique identifier for each movie.
    cast - The name of lead and supporting actors.
    title - The name of the movie
    crew - The name of Director, Editor, Composer, Writer etc.

The second dataset has the following features:-

    budget - The budget in which the movie was made.
    genre - The genre of the movie, Action, Comedy ,Thriller etc.
    homepage - A link to the homepage of the movie.
    id - This is infact the movie_id as in the first dataset.
    keywords - The keywords or tags related to the movie.
    original_language - The language in which the movie was made.
    original_title - The title of the movie before translation or adaptation.
    overview - A brief description of the movie.
    popularity - A numeric quantity specifying the movie popularity.
    production_companies - The production house of the movie.
    production_countries - The country in which it was produced.
    release_date - The date on which it was released.
    revenue - The worldwide revenue generated by the movie.
    runtime - The running time of the movie in minutes.
    status - "Released" or "Rumored".
    tagline - Movie's tagline.
    title - Title of the movie.
    vote_average - average ratings the movie recieved.
    vote_count - the count of votes recieved.


#### Let's join the two dataset on the 'id' column

In [2]:
df1.head()

movie_id                                     title  \
0     19995                                    Avatar   
1       285  Pirates of the Caribbean: At World's End   
2    206647                                   Spectre   
3     49026                     The Dark Knight Rises   
4     49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [3]:
df2.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

In [4]:
df1.rename(columns={'movie_id':'id'},inplace=True)

In [5]:
df1.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                                cast  \
0  [{"cast_id": 242, "character": "Jake Sully", "...   
1  [{"cast_id": 4, "character": "Captain Jack Spa...   
2  [{"cast_id": 1, "character": "James Bond", "cr...   
3  [{"cast_id": 2, "character": "Bruce Wayne / Ba...   
4  [{"cast_id": 5, "character": "John Carter", "c...   

                                                crew  
0  [{"credit_id": "52fe48009251416c750aca23", "de...  
1  [{"credit_id": "52fe4232c3a36847f800b579", "de...  
2  [{"credit_id": "54805967c3a36829b5002c41", "de...  
3  [{"credit_id": "52fe4781c3a36847f81398c3", "de...  
4  [{"credit_id": "52fe479ac3a36847f813eaa3", "de...

In [6]:
df2=df2.merge(df1,on='id')

In [7]:
df2.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                         ...                         runtime  \
0                        ...                           162.0   
1                        ...                           169.0   
2                        ...                           148.0   
3                        ...                           165.0   
4                        ...                           132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                    title_x vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                    title_y  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4          

#### Demographic Filtering 

Before getting started with this -

    1)we need a metric to score or rate movie
    2)Calculate the score for every movie
    3)Sort the scores and recommend the best rated movie to the users.

We can use the average ratings of the movie as the score but using this won't be fair enough since a movie with 8.9 average rating and only 3 votes cannot be considered better than the movie with 7.8 as as average rating but 40 votes. So, I'll be using IMDB's weighted rating (wr) which is given as :-

Weighted Rating(WR) = $\ ((v /(v+m))*R)+((m /(v+m))*C) $

where,

    v is the number of votes for the movie;
    m is the minimum votes required to be listed in the chart;
    R is the average rating of the movie; And
    C is the mean vote across the whole report

We already have v(vote_count) and R (vote_average) and C can be calculated as

In [8]:
C=df2['vote_average'].mean()
C

6.092171559442011

So, the mean rating for all the movies is approx 6 on a scale of 10.

The next step is to determine an appropriate value for m, the minimum votes required to be listed in the chart. 

We will use 90th percentile as our cutoff. In other words, for a movie to feature in the charts, it must have more votes than at least 90% of the movies in the list.

In [9]:
m=df2['vote_count'].quantile(0.9)
m

1838.4000000000015

Thus a movie need atleast 1838 votes to be listed in output

In [10]:
q_movies=df2.copy().loc[df2['vote_count']>=m]
q_movies.shape

(481, 23)

We see that there are 483 values which are eligible to be added in the list. Now we need to define a metric i.e a score to each movie so that it can be added to list accordingly. So we would define a function weighted_average() that calculate the score for each movie.

In [11]:
def weighted_average(x,m=m,C=C):
    v=x['vote_count']
    R=x['vote_average']
    return ((v/(v+m))*R)+((m/(v+m))*C)

In [12]:
q_movies.head()

budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                         ...                         runtime  \
0                        ...                           162.0   
1                        ...                           169.0   
2                        ...                           148.0   
3                        ...                           165.0   
4                        ...                           132.0   

                                    spoken_languages    status  \
0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1           [{"iso_639_1": "en", "name": "English"}]  Released   
2  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3           [{"iso_639_1": "en", "name": "English"}]  Released   
4           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                    title_x vote_average vote_count  \
0                                    Avatar          7.2      11800   
1  Pirates of the Caribbean: At World's End          6.9       4500   
2                                   Spectre          6.3       4466   
3                     The Dark Knight Rises          7.6       9106   
4                               John Carter          6.1       2124   

                                    title_y  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4          

In [13]:
q_movies['score']=q_movies.apply(weighted_average,axis=1)

Finally, let's sort the DataFrame based on the score feature and output the title, vote count, vote average and weighted rating or score of the top 10 movies.

In [14]:
q_movies=q_movies.sort_values('score',ascending=False)
q_movies

budget                                             genres  \
1881   25000000  [{"id": 18, "name": "Drama"}, {"id": 80, "name...   
662    63000000                      [{"id": 18, "name": "Drama"}]   
65    185000000  [{"id": 18, "name": "Drama"}, {"id": 28, "name...   
3232    8000000  [{"id": 53, "name": "Thriller"}, {"id": 80, "n...   
96    160000000  [{"id": 28, "name": "Action"}, {"id": 53, "nam...   
3337    6000000  [{"id": 18, "name": "Drama"}, {"id": 80, "name...   
95    165000000  [{"id": 12, "name": "Adventure"}, {"id": 18, "...   
809    55000000  [{"id": 35, "name": "Comedy"}, {"id": 18, "nam...   
329    94000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
1990   18000000  [{"id": 12, "name": "Adventure"}, {"id": 28, "...   
262    93000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2912   11000000  [{"id": 12, "name": "Adventure"}, {"id": 28, "...   
1818   22000000  [{"id": 18, "name": "Drama"}, {"id": 36, "name...   
3865    3300000                      [{"id": 18, "name": "Drama"}]   
330    79000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
1553   33000000  [{"id": 80, "name": "Crime"}, {"id": 9648, "na...   
94    170000000  [{"id": 28, "name": "Action"}, {"id": 878, "na...   
634    63000000  [{"id": 28, "name": "Action"}, {"id": 878, "na...   
2294   15000000  [{"id": 14, "name": "Fantasy"}, {"id": 12, "na...   
77    175000000  [{"id": 18, "name": "Drama"}, {"id": 35, "name...   
2285   19000000  [{"id": 12, "name": "Adventure"}, {"id": 35, "...   
690    60000000  [{"id": 14, "name": "Fantasy"}, {"id": 18, "na...   
287   100000000  [{"id": 18, "name": "Drama"}, {"id": 37, "name...   
2522   14000000  [{"id": 36, "name": "History"}, {"id": 18, "na...   
2731   13000000  [{"id": 18, "name": "Drama"}, {"id": 80, "name...   
494    45000000  [{"id": 10751, "name": "Family"}, {"id": 16, "...   
2091   19000000  [{"id": 80, "name": "Crime"}, {"id": 18, "name...   
298   100000000  [{"id": 80, "name": "Crime"}, {"id": 18, "name...   
571    70000000  [{"id": 18, "name": "Drama"}, {"id": 28, "name...   
3573    9000000  [{"id": 9648, "name": "Mystery"}, {"id": 53, "...   
...         ...                                                ...   
334    93000000  [{"id": 12, "name": "Adventure"}, {"id": 28, "...   
332   115000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
133   150000000  [{"id": 35, "name": "Comedy"}, {"id": 14, "nam...   
224   120000000  [{"id": 28, "name": "Action"}, {"id": 878, "na...   
53    185000000  [{"id": 12, "name": "Adventure"}, {"id": 28, "...   
937    50000000  [{"id": 14, "name": "Fantasy"}, {"id": 27, "na...   
157   140000000  [{"id": 12, "name": "Adventure"}, {"id": 18, "...   
1326   38000000  [{"id": 12, "name": "Adventure"}, {"id": 878, ...   
76    175000000  [{"id": 12, "name": "Adventure"}, {"id": 28, "...   
37    200000000  [{"id": 14, "name": "Fantasy"}, {"id": 12, "na...   
206   125000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
1682   19000000  [{"id": 12, "name": "Adventure"}, {"id": 16, "...   
189   125000000  [{"id": 18, "name": "Drama"}, {"id": 12, "name...   
898    50000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
362    88000000  [{"id": 28, "name": "Action"}, {"id": 35, "nam...   
9     250000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
27    209000000  [{"id": 53, "name": "Thriller"}, {"id": 28, "n...   
62    180000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
48    195000000  [{"id": 28, "name": "Action"}, {"id": 10751, "...   
59    200000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
215   130000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
148   144000000  [{"id": 28, "name": "Action"}, {"id": 14, "nam...   
166   130000000  [{"id": 12, "name": "Adventure"}, {"id": 28, "...   
61    176000003  [{"id": 878, "name": "Science Fiction"}, {"id"...   
1154   40000000  [{"id": 18, "name": "Drama"}, {"id": 10749, "n...   
41    200000000  [{

In [15]:
q_movies[['original_title','vote_count','vote_average','score']].head(15)

original_title  vote_count  \
1881                           The Shawshank Redemption        8205   
662                                          Fight Club        9413   
65                                      The Dark Knight       12002   
3232                                       Pulp Fiction        8428   
96                                            Inception       13752   
3337                                      The Godfather        5893   
95                                         Interstellar       10867   
809                                        Forrest Gump        7927   
329       The Lord of the Rings: The Return of the King        8064   
1990                            The Empire Strikes Back        5879   
262   The Lord of the Rings: The Fellowship of the Ring        8705   
2912                                          Star Wars        6624   
1818                                   Schindler's List        4329   
3865                                           Whiplash        4254   
330               The Lord of the Rings: The Two Towers        7487   

      vote_average     score  
1881           8.5  8.059258  
662            8.3  7.939256  
65             8.2  7.920020  
3232           8.3  7.904645  
96             8.1  7.863239  
3337           8.4  7.851236  
95             8.1  7.809479  
809            8.2  7.803188  
329            8.1  7.727243  
1990           8.2  7.697884  
262            8.0  7.667341  
2912           8.1  7.663813  
1818           8.3  7.641883  
3865           8.3  7.633781  
330            8.0  7.623893

Thus we are now able to define a demographic filter that provides a general and a primitive way of providing user with a movie list that isn't user specific but does considers that the movie that appeal to a more users might appeal the new user.

## Content Based Filtering

In this recommender system the content of the movie (overview, cast, crew, keyword, tagline etc) is used to find its similarity with other movies. Then the movies that are most likely to be similar are recommended.

#### Plot description based Recommender¶

We will compute a pair wise similarity score for each movie and then suggest a movie based on the score. The movie that has highest similarity score is recommended.

In [16]:
df2['overview'].head()

0    In the 22nd century, a paraplegic Marine is di...
1    Captain Barbossa, long believed to be dead, ha...
2    A cryptic message from Bond’s past sends him o...
3    Following the death of District Attorney Harve...
4    John Carter is a war-weary, former military ca...
Name: overview, dtype: object

 Now we'll compute Term Frequency-Inverse Document Frequency (TF-IDF) vectors for each overview.

Now if you are wondering what is term frequency , it is the relative frequency of a word in a document and is given as (term instances/total instances). Inverse Document Frequency is the relative count of documents containing the term is given as log(number of documents/documents with term) The overall importance of each word to the documents in which they appear is equal to TF * IDF

Thus we will get a matrix where each row is denoted by a word in the overviews of all the movies and the column is used to denote a vector for each overview. Thus a given mat[][] will denote the frequency of a word in a given document(overview)

In [22]:
# import TfidfVectorizer from skelarn
from sklearn.feature_extraction.text import TfidfVectorizer

# initialize an TfidfVectorizer object. Remove all words like 'the', 'a' , 'an'
tfidf=TfidfVectorizer(stop_words='english')

#replace NaN values with ''
df2['overview']=df2['overview'].fillna('')

# create a tfidf matrix
tfidf_matrix=tfidf.fit_transform(df2['overview'])

# size of the matrix
tfidf_matrix.shape

(4803, 20978)

We see that over 20,000 different words were used to describe the 4800 movies in our dataset.

We use the cosine similarity score since it is independent of magnitude and is relatively easy and fast to calculate. 

Since we have used the TF-IDF vectorizer, calculating the dot product will directly give us the cosine similarity score. Therefore, we will use sklearn's linear_kernel() instead of cosine_similarities() since it is faster.

In [23]:
# Import linear_kernel
from sklearn.metrics.pairwise import linear_kernel

# compute the cosine similarity matrix
cosine_sim = linear_kernel(tfidf_matrix,tfidf_matrix)

In [24]:
cosine_sim

array([[1.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ],
       [0.        , 1.        , 0.        , ..., 0.02160533, 0.        ,
        0.        ],
       [0.        , 0.        , 1.        , ..., 0.01488159, 0.        ,
        0.        ],
       ...,
       [0.        , 0.02160533, 0.01488159, ..., 1.        , 0.01609091,
        0.00701914],
       [0.        , 0.        , 0.        , ..., 0.01609091, 1.        ,
        0.01171696],
       [0.        , 0.        , 0.        , ..., 0.00701914, 0.01171696,
        1.        ]])

We are going to define a function that takes in a movie title as an input and outputs a list of the 10 most similar movies. Firstly, for this, we need a reverse mapping of movie titles and DataFrame indices. In other words, we need a mechanism to identify the index of a movie in our metadata DataFrame, given its title.

In [27]:
indices=pd.Series(df2.index,index=df2['original_title']).drop_duplicates()

In [28]:
indices

original_title
Avatar                                            0
Pirates of the Caribbean: At World's End          1
Spectre                                           2
The Dark Knight Rises                             3
John Carter                                       4
Spider-Man 3                                      5
Tangled                                           6
Avengers: Age of Ultron                           7
Harry Potter and the Half-Blood Prince            8
Batman v Superman: Dawn of Justice                9
Superman Returns                                 10
Quantum of Solace                                11
Pirates of the Caribbean: Dead Man's Chest       12
The Lone Ranger                                  13
Man of Steel                                     14
The Chronicles of Narnia: Prince Caspian         15
The Avengers                                     16
Pirates of the Caribbean: On Stranger Tides      17
Men in Black 3                                   

In [41]:
def get_recommendations(title,cosine_sim=cosine_sim):
    # get the index of the movie from it's title
    idx=indices[title]
    
    # get the pair wise similarity score with the given movie
    sim_scores=list(enumerate(cosine_sim[idx]))
    
    #sort in reverse order
    sim_scores=sorted(sim_scores,key= lambda x:x[1],reverse=True)
    
    #get top 10 movies
    sim_scores=sim_scores[1:11]
    
    #get the movie indices
    movie_indices=[i[0] for i in sim_scores]
    
    #return the titles
    return df2['original_title'].iloc[movie_indices]
    

In [42]:
get_recommendations('The Dark Knight Rises')

65                              The Dark Knight
299                              Batman Forever
428                              Batman Returns
1359                                     Batman
3854    Batman: The Dark Knight Returns, Part 2
119                               Batman Begins
2507                                  Slow Burn
9            Batman v Superman: Dawn of Justice
1181                                        JFK
210                              Batman & Robin
Name: original_title, dtype: object

In [46]:
get_recommendations("The Amazing Spider-Man")

38       The Amazing Spider-Man 2
159                    Spider-Man
5                    Spider-Man 3
1572    Forgetting Sarah Marshall
30                   Spider-Man 2
572                          Hook
143                           Pan
4246                      Trucker
1179              I Love You, Man
641                      Due Date
Name: original_title, dtype: object

In [55]:
get_recommendations("Lost in Space")

1709                        キャプテンハーロック
432                        Deep Impact
56                    Star Trek Beyond
1914                         Lifeforce
373                    Mission to Mars
1352                           Gattaca
3184                   The Ice Pirates
52      Transformers: Dark of the Moon
1294                          Serenity
1531                         Moonraker
Name: original_title, dtype: object

This recommendation system is just based on the movie overview and does not considers any parameter like cast, crew, keywords and overview . Now we are gonna make a system that takes top 3 crew members, directos, keywords and overview into consideration to provide a much better recommendation